### ISSUE TITLE
robot reason concept of new axioms needs to be changed

### ISSUE THREAD

**Issue by cmungall** [ontodev/robot#518]
Currently in reason we have `-a` for annotating inferred axioms, and `-n` for placing these in a new ontology.

I am responsible for these, and I under-specified these. I don't think the current behavior is useful. The intent was for both of these options to produce information on *new* axioms. With the current behavior `-a` causes *every* subClassOf to be stamped, and `-n` causes *every* axiom to be placed in a new ontology. Trivially, every axiom entails itself but this is not useful!

For example, given this an input:

```
[Term]
id: X:1
is_a: X:2
```

running reason with `-a true` gives

```
[Term]
id: X:1
is_a: X:2 {is_inferred="true"}
```

(using obo format here as it's the most compact and readable way to write simple annotated axioms but the format is irrelevant).

I think even though the docs are underspecified, this can be treated as a bug, and we can go ahead and fix and not worry about compatibility. There is no way anyone could have been *meaningfully* relying on current behavior.

I could have sworn the behavior *used* to be different, perhaps some change in the reason code had the effect on changing this, but it's on me for not producing sufficient tests.

Note: I believe @dosumis may have been confused about the behavior here: #291

**Comment by jamesaoverton** [comment]
I agree that this sounds like a bug. There are some unit tests for this functionality, so it would be good to know why the bad behaviour is allowed.

**Comment by balhoff** [comment]
Possibly related: when ontology A imports ontology B, I find that ROBOT inserts inferred subclass axioms into ontology A, even if they already are asserted in ontology B. I wouldn't think this is intended.

**Comment by jamesaoverton** [comment]
That also sounds like a bug.

It would have been nice to get all this right the first time around, years ago. If it’s too badly broken now we can always create a new `reason2` command or something.

**Comment by balhoff** [comment]
@jamesaoverton this was my mistake: I somehow missed the`--exclude-duplicate-axioms` option. I think the default should be `true` instead of `false`, but at least it's there!

**Comment by balhoff** [comment]
I've been digging into ReasonOperation. My issue and @cmungall's issue are actually related. It turns out that when the OWL API outputs inferred axioms using `InferredOntologyGenerator`, it only checks the destination ontology you pass to see whether an axiom is new. I always thought that it would instead check the root ontology used by the reasoner. So if you pass a fresh ontology, so that you can segregate inferred axioms, it will put (for example) ALL subclass axioms, not just those not asserted in the ontology used by the reasoner. ROBOT always first places inferred axioms in a new ontology, and later chooses whether to empty the original before adding these in or not. So OWL API is always comparing to an empty ontology, regardless.

So you essentially always need to use `--exclude-duplicate-axioms`. In this case, tagging inferred axioms works as one would expect. I can't think of a case when you would not want to exclude duplicate axioms. When merging inferences in, it just seems to result in copying axioms from imports into the main ontology.

Currently ROBOT doesn't ignore axiom annotations when comparing axioms for excluding duplicates. It probably should.

**Comment by balhoff** [comment]
My feeling is that if we want to "fix" @cmungall's issue, we need to get rid of the `--exclude-duplicate-axioms` option. The operation needs to know which axioms are "inferred", which to me (in the case of subclass axioms) would be any subclass relationships found by the reasoner which aren't present as asserted axioms in the ontology or its imports. If the user chooses `--create-new-ontology true`, these are the axioms that would be put in the new ontology. The only way to obtain a set of axioms to be considered "inferred" is to do the comparison which `--exclude-duplicate-axioms` does (so I think it should always be `true`).

@cmungall @jamesaoverton @rctauber @matentzn can you think of a sort of inference which we would classify as a "inferred axiom" but which would be a duplicate? I think if it is a duplicate then I wouldn't call it an inferred axiom. Otherwise are we saying that if you output inferred subclass axioms into a new ontology (with `--include-indirect false`), this is the complete nonredundant subclass tree, not just new relationships? I guess I could live with that; if so I will spell it out better in the documentation. Thoughts?

**Comment by matentzn** [comment]
I am a bit scared of this kind of discussion, as fundamentally, there is no really good distinction between inferred and asserted that is complete in any sense of the word. There are trivial inferences that may be inferred yet duplicate.

Simple example (O1):

```
A sub B and C
```

ELK implies:

```
ax1: A sub B
ax2: A sub C
```
Are they are now asserted or inferred?

Other a less controversial example (O2):

```
ax2.1: A sub B [dbxref: xyz]
```

Elk implies:

```
ax2.2.: A sub B
```

Here it seems of course that  ax2.2 should not be labelled inferred.

What about tautologies (axioms that always follow from the ontology)?

```
ax3.1: A sub owl:Thing
```

What about equivalences (O4)?

```
ax4.1: A = B
```

And Elk infers 
```
ax4.2: A sub B
```

I guess if you restrict yourself to the case of:

**Transitive reduct, non-tautological subsumptions between classnames without axiom annotations** @balhoff solution is probably the right one.

**Comment by jamesaoverton** [comment]
@matentzn is right to point out that there's a lot of subtleties here, but I think @balhoff is heading in the right direction. "Inferred" axioms:

- are found by the reasoner
- are not asserted in the ontology or its imports -- ignoring annotations
- are not tautologous -- this just adds noise
- may be trivial -- we need to load the results into dumb systems (that don't do any inference of their own, such as a basic triplestore) and find the inferred stuff
